### ISSUE TITLE
Improve (Taproot) signing API

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1470]
I have a fresh experience trying to implement some Taproot-based contract and there are a few pain points, among things like `Script` not  being tagged:

* Types in `taproot` module are mostly good (there are other issues open for not that good parts) but it'd help having an example that doesn't over-complicate things going through PSBTs.
* The contents of `Witness` is not obvious to someone who didn't read the BIP and is not cursed with insane attention to details. A `WitnessBuilder` might help here.
* "Is just converting `TapSighashHash` (what a weird name) to `secp256k1::Message` the right thing to do?
* According to the BIP if you use non-`Default` `SighashType` the signature has to be 65 bytes - having the extra flag byte. This byte is "lost" when computing `TapSighashHash`. I think `TapSighashHash` should be internally 65 bytes (storing the extra byte) and have `to_message` method and `sign(&KeyPair)` method which outputs `TapScriptSignature` which can be added with `Witness::push_signature` method.

Notably regarding contents: it's no obvious that the signatures must be in the reverse order compared to the order of public keys in the script. I happened to guess this correctly ("It's a stack of course") but I could totally imagine someone missing this. I don't expect type system to be able to help with this so just a note in the `WitnessBuilder` doc would be fine.

And we also need more good examples. There's a trick that allows one adding `documentation` module to the doc that doesn't show up in real code. It feels slightly weird but I'd be totally OK with this.

**Comment by apoelstra** [comment]
concept ACK having a `WitnessBuilder` of some sort, although I imagine we'll want to bicker a bit about how general it should be. You mention, for example, that when doing multisig with `CHECKSIGADD` you need to reverse keys wrt signatures, and that this is surprising. True, and true, but I consider this sort of thing to be the job of rust-miniscript.

Though perhaps for "simple multisigs" we should add direct support here, since it's a common case, and in Taproot we can get a lot of mileage by supporting common cases, since users can hide the uncommon ones in unused tapbranches where our code won't trip over them.

The lost sighash flag byte sounds like it's unambiguously a bug we need to fix.

**Comment by Kixunil** [comment]
> I consider this sort of thing to be the job of rust-miniscript.
>
> Though perhaps for "simple multisigs" we should add direct support here, since it's a common case

Yeah, my thinking was exactly "this is simple, I shouldn't need yet another library for this". In my case the multisig is literally hard-coded (statically known), only the keys are changed.

> sounds like it's unambiguously a bug we need to fix.

Well, it's kinda understandable the way it works since the `SighashCache` methods only return the hash which needs to be signed (potentially by an external signer) and then signature goes into `Witness` not the hash.

And frankly it didn't occur to me to search for signature serialization support now I see we already have. If there was a method on `TapSighashHash` producing it (with a supplied key) it'd be easier to find.
